### PR TITLE
[ASTBridging] Avoid temporary SmallVector for 'InheritedEntry' list

### PR DESCRIPTION
**Pull request by rintaro**
We know the number of the entries. Allocate the storage in ASTContext directly, then initialize it transforming the bridged values. Add ASTContext::AllocateTransform() helper function.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/AST/ASTBridging.cpp [hunk 1]
```diff
@@ -1133,8 +1130,8 @@ BridgedBraceStmt BridgedBraceStmt_createParsed(BridgedASTContext cContext,
   }
 
   ASTContext &context = cContext.unbridged();
-  return BraceStmt::create(context, cLBLoc.unbridged(),
-                           context.AllocateCopy(nodes), cRBLoc.unbridged());
+  return BraceStmt::create(context, cLBLoc.unbridged(), nodes,
```
Comment (rintaro) [inline lib/AST/ASTBridging.cpp:1133]: `nodes` are copied in `BraceStmt::create` anyway.

File: include/swift/AST/ASTContext.h [hunk 2]
```diff
@@ -521,6 +521,18 @@ class ASTContext final {
                               setVector.size());
   }
 
+  template <typename Output, typename Range>
+  ArrayRef<Output> AllocateTransform(
+      Range &&input,
+      llvm::function_ref<Output(const decltype(input.front()) &)> transform,
```
Comment (AnthonyLatsis) [inline include/swift/AST/ASTContext.h:527]: Is the transform param type intentionally not `const Range::value_type &` or `Range::const_reference`?

**Review by AnthonyLatsis — COMMENTED** [review]
Nice!

**Review by hamishknight — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**